### PR TITLE
🔀🐛✨📝 Merging PR #51 — Course 1 Exercises (Part 2, in progress) 

### PR DESCRIPTION
**Pull request by heng-yuwen**
# ✈  [**In Progress**] Merging PR #51 — Merging changes to Course 1 - Computer Vision [Exercises](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/tree/main/1-Computer-Vision/Exercises).

Includes bug fixes and API update to [TensorFlow v2.13.0](https://www.tensorflow.org/api_docs). To view the diffs for `.ipynb` files, use a Jupyter notebook diff renderer, e.g., [ReviewNB](https://app.reviewnb.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/51/).

  

**Please check if the PR fulfills these requirements**:
- [x] The commit message follows repository guidelines;
- [x] Tests for the changes have been added (for bug fixes / features);
- [x] Docs have been added / updated (for bug fixes / features).

## :label: Type of change(s)

- [x] :bug: Bug fix (non-breaking change which fixes an issue);
- [x] :sparkles: New feature (non-breaking change which adds functionality);
- [x] :memo: This change requires a documentation update.

 

## 🧪 Testing and verification
Modifications have been tested / verified against known solutions (i.e., 'ground truth') by comparing updated cell outputs to expected values provided by Udacity.

In [`../2022-08-27-Logistic-Regression.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-3-1-Logistic-Regression/2022-08-27-Logistic-Regression.ipynb):

 - [x] :white_check_mark: `tnp.experimental_enable_numpy_behavior()`: no longer raising `Not creating XLA devices, tf_xla_enable_xla_devices not set` error when calling `tnp`;
 - [x] :white_check_mark: Use of TensorFlow v2.13.0 with `tnp.experimental_enable_numpy_behavior()` does not affect Tensor output values produced:
    - [x] :white_check_mark: `check_softmax`: passing test case — returns softmax calculation within range of expected value;
    - [x] :white_check_mark: `check_ce`: passing test case — returns cross-entropy calculation within range of expected value; 
    - [x] :white_check_mark: `check_model`: passing test case — logistic regression `model` returns expected output shape;
    - [x] :white_check_mark: `check_acc`: passing test case — `accuracy` scoring function returns expected accuracy value.
 
In [`../2022-08-29-Stochastic-Gradient-Descent.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb):

 - [x] :white_check_mark: `mean_acc = tf.math.reduce_mean(tf.concat(accuracies, axis=0))`: returns expected behaviour when removing `tf.concat` (NOTE: `tf.concat` is not allowed for scalars, use `tf.stack` or directly calculate mean with the list);
 - [x] :white_check_mark: SGD algorithm returns expected values after migration to TensorFlow v2.13.0:
    - [x] :white_check_mark: SGD algorithm achieves similar accuracy during inference run over GTSRB dataset — determined visually by comparing the accuracy curve;
    - [x] :white_check_mark: SGD algorithm achieves similar cross-entropy loss during inference run over GTSRB dataset — determined visually by comparing the cross-entropy loss curve.
 

## 💬 More info

The following notebook files have been added / modified:
* [`../2022-08-27-Logistic-Regression.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-3-1-Logistic-Regression/2022-08-27-Logistic-Regression.ipynb):
   * `tnp.experimental_enable_numpy_behavior()`: enables support of NumPy behaviour with experimental backend for TensorFlow;
      * _"This call enables type promotion in TensorFlow and also changes type inference, when converting literals to tensors, to more strictly follow the NumPy standard."_ — [NumPy API on TensorFlow | TensorFlow Core Guide](https://www.tensorflow.org/guide/tf_numpy);
   * Update cross-entropy function by removing the $C$ in the denominator (as written by E. Bendersky in [this](https://eli.thegreenplace.net/2016/the-softmax-function-and-its-derivative/) article).

$$
\begin{align}
    \phi_{softmax}\left(z_{i}\right) 
    = \frac{C\mathcal{e}^{z_{i}}}{\sum_{j=0}^{k} C\mathcal{e}^{z_{k}^{i}}} = \frac{\mathcal{e}^{z_{i} + \mathrm{log}\left(C\right)}}{\sum_{j=0}^{k} \mathcal{e}^{z_{k}^{i} + \mathrm{log}\left(C\right)}}.
    \end{align}
$$

* [`../2022-08-29-Stochastic-Gradient-Descent.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-3-2-Stochastic-Gradient-Descent/2022-08-29-Stochastic-Gradient-Descent.ipynb):
   * Set `ENV_COLAB` to False by default like other notebooks;
   * Add `-o` flag to unzip command;
   * [Fix error caused by applying concat to scalar list](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/51/commits/eb2d1ce94ffd341e2610c63b803cf3d937a85ae0);
   * Explicitly mention L2 loss in gradient formula;
   * [Correct the tensor dimension of `y_hat` defined in `accuracy` function comments.](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/51/commits/207044e0316a2deb70e201575a7d374eab325eea)

### CODE REVIEW

**Review by jonathanloganmoran — COMMENTED** [review]
Hello @heng-yuwen; I have reviewed the changes made here and can proceed with an _approval_ based on the latest results of the test cases and expected outputs from these two notebook files. 

Thank you for the detailed [💬 More info](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/pull/51#:speech_balloon:-More-info) section—I see your recent additions match entirely what I was preparing to mention in your PR memo. I will also be adding the passing test cases in your PR memo as the sanity check for this PR, as I am unable to reproduce the TF tests myself (I am without access to a GPU at the moment).

**UPDATE 2023-08-16 11:33 PT**: My changes to PR #51 memo have been finalised. Before I push this review, I am seeing that the latest stable version of TensorFlow API is v2.12, are we sure that the migration to v2.13.0 is necessary to achieve our desired functionality?

**Comment by heng-yuwen** [comment]
Hi @jonathanloganmoran, thanks for organising the memo. I think v2.13 is the stable version since it is now the default one installed following the official documentation. Also, I think the content included in this project will not break even if some of the latest feature is not functional since these new features are not used. Moreover, v2.13 fixed a bug of tf.Variable, which is widely used in this project.

**Comment by jonathanloganmoran** [comment]
> Hi @jonathanloganmoran, thanks for organising the memo. I think v2.13 is the stable version since it is now the default one installed following the official documentation. Also, I think the content included in this project will not break even if some of the latest feature is not functional since these new features are not used. Moreover, v2.13 fixed a bug of `tf.Variable`, which is widely used in this project.

Great to know, thank you for your justification. 

I will proceed with merging PR #51 into the working branch [`jonathanloganmoran:pr49-course-1-revision-1-heng-yuwen`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/tree/pr49-course-1-revision-1-heng-yuwen). I should be able to merge into `main` afterwards without concern, however, I would like to warn you that the use of TensorFlow v2.13.0 _might_ break functionality on any notebooks also making use of the [Waymo Open Dataset](https://github.com/waymo-research/waymo-open-dataset) and [TensorFlow Object Detection](https://github.com/tensorflow/models/tree/master/research/object_detection) APIs (e.g., [`2022-08-03-Creating-TF-Records.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-1-3-Creating-TF-Records/2022-08-03-Creating-TF-Records.ipynb)). IIRC, you must specifically downgrade TF to a compatible version when using these two APIs. Nonetheless, I believe the two notebooks affected in this PR will not present us with any new issues related to this upgrade.

**Comment by heng-yuwen** [comment]
> > Hi @jonathanloganmoran, thanks for organising the memo. I think v2.13 is the stable version since it is now the default one installed following the official documentation. Also, I think the content included in this project will not break even if some of the latest feature is not functional since these new features are not used. Moreover, v2.13 fixed a bug of `tf.Variable`, which is widely used in this project.
> 
> Great to know, thank you for your justification.
> 
> I will proceed with merging PR #51 into the working branch [`jonathanloganmoran:pr49-course-1-revision-1-heng-yuwen`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/tree/pr49-course-1-revision-1-heng-yuwen). I should be able to merge into `main` afterwards without concern, however, I would like to warn you that the use of TensorFlow v2.13.0 _might_ break functionality on any notebooks also making use of the [Waymo Open Dataset](https://github.com/waymo-research/waymo-open-dataset) and [TensorFlow Object Detection](https://github.com/tensorflow/models/tree/master/research/object_detection) APIs (e.g., [`2022-08-03-Creating-TF-Records.ipynb`](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-1-3-Creating-TF-Records/2022-08-03-Creating-TF-Records.ipynb)). IIRC, you must specifically downgrade TF to a compatible version when using these two APIs. Nonetheless, I believe the two notebooks affected in this PR will not present us with any new issues related to this upgrade.

As for the notebook [2022-08-03-Creating-TF-Records.ipynb](https://github.com/jonathanloganmoran/ND0013-Self-Driving-Car-Engineer/blob/main/1-Computer-Vision/Exercises/1-1-3-Creating-TF-Records/2022-08-03-Creating-TF-Records.ipynb), I will go through it in the next few days. I am setting up my computing server with the Waymo dataset. It should not be difficult to update its API usage as well.

**Comment by jonathanloganmoran** [comment]
@heng-yuwen: That would be awesome if you could give that notebook a test. Since I am currently without access to a Linux workstation, I am unable to test the Waymo Open Dataset API. As far as using a TF v2.13.0 backbone for the TF2 Object Detection API is concerned, that is something I should be able to use conjure Colab for. However, there was never a shortage of problems created with that approach in the past 😉  Please let me know how it goes for you in your experiments. In my past work, I did my best to modify the standard-issue 'tutorial' notebooks from Waymo and TF2OD teams to run successfully for the TF version I was using.

**Comment by heng-yuwen** [comment]
> 

I think I am aware of your concern. Yes, upgrading the TensorFlow version may break the code. I will make environment requirements.txt in my next PR and run all the code in my cloud server to validate the API.
